### PR TITLE
TEAM GLITZY- Tanja, Jan, Severin, and Amy

### PR DESCRIPTION
**Pull request by ayjlee**
# bEtsy
Congratulations! You're submitting your assignment! These comprehension questions should be answered by all members of your team, not by a single teammate.

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| How did your team break up the work to be done? |  Task leads came up with work plan, delegated assignments. We group planned on design of MVC and database schema, and had a week 1 goal of completing functionality and tests for Guests, and week 2 goal for finishing Merchant functionality and tests. Tasks were divided into pair programming or solo assignments depending on difficulty, and interconnectedness, as well as individual strengths and goals.  |
| How did your team utilize git to collaborate? |  We primarily used git branches and pull requests.  |
| What did your group do to try to keep your code DRY while many people collaborated on it?  |  We left time at the end to identify as a group areas for refactoring, then divide up refactoring tasks to DRY up code.  |
| What was a technical challenge that you faced as a group?  |  Deploying to Heroku utilizing Paperclip and AWS authentication keys, checking for dependencies between gem versions before deploying, handling mismatches between local schema and Heroku schema in case of dropped migrations.   |
| What was a team/personal challenge that you faced as a group?  |  At various points, team members faced frustrating episodes of learning how to use git better, the team rallied together to make sure that no one was left behind because everyone having a functional version of the repository was essential to moving forward.  |
| What could your team have done better?  |  As a group, coming to a consensus about overall design and wireframes earlier in the process to avoid ambiguity and empower people to move forward style decisions based on the frameworks given.  |
| What was your application's ERD? (include a link) |  https://github.com/ayjlee/betsy/blob/master/erd.pdf  |
| What is your Trello URL? |   https://trello.com/b/8aCdTueZ/glitzy     |
| What is the Heroku URL of your deployed application? |  https://glitzy-store.herokuapp.com      |

### CODE REVIEW

File: test/models/category_test.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    it "has products" do
+      food_category.must_respond_to :products
+    end
+
+    it "can have many products" do
+      food_category.products << products(:chocolate)
+      food_category.products.first.must_be_kind_of Product
+
+      food_category.products << products(:tears)
+      food_category.products.count.must_equal 2
```
Comment (CheezItMan) [inline test/models/category_test.rb:17]: you should have a `food_category.save` here ensuring that no validations etc get broken.

File: test/models/order_item_test.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  describe 'validations' do
+    it "requires a quantity" do
+      item_no_quantity = OrderItem.new
+      item_no_quantity.valid?.must_equal false
+      item_no_quantity.errors.messages.must_include :quantity
+    end
+
+    it "requires that quantity is a positive integer " do
+      item1.quantity.must_be_kind_of Integer
+      item1.quantity.must_be :>, 0
```
Comment (CheezItMan) [inline test/models/order_item_test.rb:31]: These two seem redundant.

File: test/models/order_item_test.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+require "test_helper"
```
Comment (CheezItMan) [inline test/models/order_item_test.rb:1]: No test for `available?`  ???

Remember to test custom methods.

File: test/models/order_test.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      order1.must_be_kind_of Order
+      order1.order_items.each do |item|
+        item.must_be_kind_of OrderItem
+      end
+    end
+  end
+
+  describe "validations" do
+
+    it "an order requires a status" do
```
Comment (CheezItMan) [inline test/models/order_test.rb:25]: I would also test if `status` is `nil`.

File: test/models/order_test.rb [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+      test_total = 0
+      order2.order_items.each do |order|
+        test_total += (order.product.price * order.quantity)
+      end
+      order2.total_cost.must_equal test_total
+    end
+  end
+
+  describe "custom model methods" do
```
Comment (CheezItMan) [inline test/models/order_test.rb:75]: Shouldn't the `total_cost` method be tested in here as opposed to outside?

File: test/models/order_test.rb [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    let(:bubbles) {users(:bubbles)}
+    let(:user) {users(:tanja)}
+    describe "self.by_user(user)" do
```
Comment (CheezItMan) [inline test/models/order_test.rb:78]: You should make sure it's in the proper order.

File: test/models/product_test.rb [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+
+    it "has a list (can have many) order items" do
+      product.must_respond_to :order_items
+      product.order_items.each do |item|
+        item.must_be_kind_of OrderItem
+      end
+    end
+  end
+
+  describe "validations" do
```
Comment (CheezItMan) [inline test/models/product_test.rb:34]: What about validations on the photo field?

File: test/models/product_test.rb [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+      end
+
+    end
+
+    describe "self.to_merchant_hash method" do
+      it "returns a hash with user names as keys and product arrays as values" do
+        merchant_hash = Product.to_merchant_hash
+
+        merchant_hash.must_be_kind_of Hash
+        merchant_hash.keys.count.must_equal 4
```
Comment (CheezItMan) [inline test/models/product_test.rb:149]: I would instead loop through an array of the keys and say:

`merchant_hash.keys.must_include key`

File: test/models/review_test.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
+require "test_helper"
+
+describe Review do
+  let(:review) { reviews(:one) }
+  let(:product) { products(:soap)}
+
+  describe "relations" do
```
Comment (CheezItMan) [inline test/models/review_test.rb:7]: What about a review belongs to a user?

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        newuser.must_respond_to :total_revenue_by_status
+      end
+
+      it "calculates revenue for zero value" do
+        newuser = User.new(name: "Bob")
+        newuser.total_revenue_by_status(status).must_equal 0.0
+      end
+
+      it "returns correct value for incomplete order" do
+        user.total_revenue_by_status("incomplete").must_equal pending_order_item.product.price * pending_order_item.quantity
```
Comment (CheezItMan) [inline test/models/user_test.rb:111]: need more testing for different status levels

File: test/controllers/categories_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    it " Data should reflect newly created category" do
+      proc {
+        post categories_path, params: { category: {name: "some cat"}}
+      }.must_change 'Category.count', 1
+    end
+
+
+    it " Should render template and not affect the data if failed " do
```
Comment (CheezItMan) [inline test/controllers/categories_controller_test.rb:28]: incomplete test?

File: test/controllers/categories_controller_test.rb [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      proc {
+        post categories_path, params: { category: {name: ""}}
+      }
```
Comment (CheezItMan) [inline test/controllers/categories_controller_test.rb:31]: maybe `.must_change 'Category.count', 0`

I would also suggest that this should redirect instead of render.

File: test/controllers/categories_controller_test.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
+require "test_helper"
+
+describe CategoriesController do
+  describe "when user is logged in" do
```
Comment (CheezItMan) [inline test/controllers/categories_controller_test.rb:4]: This structure is good.

File: test/controllers/order_items_controller_test.rb [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+  describe "OrderItem#new" do
+    it "works" do
+      get new_product_order_item_path(products(:tears).id)
+      must_respond_with :success
+    end
+  end
+
+  describe "OrderItem#edit" do
+    it "works" do
+      get edit_order_item_path(chocolate1.id)
```
Comment (CheezItMan) [inline test/controllers/order_items_controller_test.rb:31]: Should an anonymous user be able to edit an order_item they didn't create?

File: test/controllers/order_items_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,238 @@
...
+          order: soap2.order
+        }
+      }
+      #Action
+
+      patch order_item_path(soap2.id), params: oi_update_data
+
+      #Assert
+
+      must_respond_with :redirect
```
Comment (CheezItMan) [inline test/controllers/order_items_controller_test.rb:178]: You should also test the flash message as well.

File: test/controllers/order_items_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,238 @@
...
+    end
+
+    it "cannot delete an order_item from cart when order status is paid" do
+      #Arrange
+      soap2.order.status.must_equal "paid"
+      #Action
+      delete order_item_path(soap2.id)
+      #Assert
+      OrderItem.find_by(id: soap2.id).must_equal soap2
+      must_redirect_to order_path(paid.id)
```
Comment (CheezItMan) [inline test/controllers/order_items_controller_test.rb:212]: Also some flash error message and status code should be tested.

File: test/controllers/order_items_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+    end
+  end
+
+  describe "OrderItem#mark_shipped" do
+
+    let(:merchant) {users(:je)}
+    let(:merchant_order_item) {order_items(:orderitem4)}
+
+    it "ensures a paid order_item's default shipping status is: not shipped" do
+      login(merchant)
```
Comment (CheezItMan) [inline test/controllers/order_items_controller_test.rb:223]: Why login?

File: test/controllers/orders_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+      }
+      # order has 0 order items
+      get root_path
+      order_id = session[:order_id]
+      order = Order.find_by(id: order_id)
+      puts "ORDER HAS #{order.order_items.count}"
+      puts "order status is #{order.status}"
+      order.status.must_equal "incomplete"
+      patch place_order_path, params: order_data
+      order.status.must_equal "incomplete"
```
Comment (CheezItMan) [inline test/controllers/orders_controller_test.rb:106]: There should also be a status code and flash message to check.

**Comment by CheezItMan** [comment]
# bEtsy
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
| Appropriate Git Usage with all members contributing	| Good number of commits and a large number of commit messages.   |
| Answered comprehension questions	|  Check, I'm glad branching helped and that you pulled together to keep everyone up.  |
| Trello board is created and utilized in project management | It looks like you got most of what you wanted done!  |
| Heroku instance is online |  Check, nice |
|  **General** |   |
| Nested routes follow RESTful conventions | Good RESTful routes where appropriate.  You do have some unused routes still in your `routes.rb` file, like the users index action.    |
| OAuth used for User authentication  |  Check  |
| Functionality restricted based on user roles  | As a merchant I can, by changing the URL edit and change another merchant's product.  See [here](https://glitzy-store.herokuapp.com/products/1048).  The use of a filter would help ensure this.    |
| Products can be added and removed from cart | Check  |
| Users can view past orders | Check  |
| Merchants can add, edit and view their products | Check  |
| Errors are reported to the user  | Check  |
| **Order Functionality** |   |
| Reduces products' inventory  | Check, but I can add a product twice allowing me to check out with more item than were in inventory.  |
| Cannot order products that are out of stock  | Check  |
| Changes order state  | Check  |
| Clears current cart  |  Check |
| **Database** |   |
| ERD includes all necessary tables and relationships  | The ERD isn't using standard notation, but has the models used and their relationships.  |
| Table relationships  | Relationships present.   |
| **Models** |   |
| Validation rules for Models |  Check, I like how you used the constant to list the possible order status states. I also like how you add validations for the product images with PaperClip.   |
| Business logic is in the models | A good amount of business logic is in the models.    |
| **Controllers** |   |
| Controller Filters used to DRY up controller code |  Good use of the `render_404` method and filters like `find_user` and `current_order`.  Some more DRYing up could be done however, like in the orders controller.   |
| **Testing** |   |
| Model Testing | Lots of testing.  I left some notes in the code |
| Controller Testing | Left some notes in your code.  Good overall, you need to do more checks of status codes and flash messages.   |
| Session Testing | Good session testing!  |
| SimpleCov at 90% for controller and model tests | Almost exactly 90%  |
|  **Front-end** |   |
| The app is styled to create an attractive user interface | Very nicely done.  |
| The app layout uses Foundation's Grid | Somewhat responsive.  Some pages like fulfillment have issues with small sizes |
| Content is organized with HTML5 semantic tags | Some areas (product index and order index) still have divs where semantic tags would do better. |
| CSS is DRY | Some DRYing could be done, color layouts etc. |
|  **Overall** |  Very nicely done.  You had almost all the functionality and a few bugs, but overall outstanding work.     |

File: test/controllers/products_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+
+      get product_path(nonexistant_id)
+      must_respond_with :not_found
+    end
+
+  end
+
+  describe "edit" do
+    it "will get the edit form for an existing product" do
+      get edit_product_path(Product.first)
```
Comment (CheezItMan) [inline test/controllers/products_controller_test.rb:89]: What about if the user isn't logged in!!!!  OMG!  Or if the user isn't the owner!

File: test/controllers/products_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,192 @@
...
+      get edit_product_path((last_product_id + 1))
+      must_respond_with :not_found
+    end
+
+  end
+
+  describe "update" do
+
+    it "updates with valid data and an existing product " do
+      product = products(:chocolate)
```
Comment (CheezItMan) [inline test/controllers/products_controller_test.rb:107]: Ditto!

**Review by CheezItMan — COMMENTED** [review]
Left you some notes, overall nicely done!
